plan/attributes.go

```diff
@@ -182,19 +195,24 @@ func CheckSuccessorsMustRequire(node *PhysicalPlanNode) error {
 // succ passes through the attribute and one of its successors requires the attribute.
 // If the attribute is not required, this function returns false and the node that neither passes
 // along nor requires the attribute.
-func requiredBySuccessor(requiredAttr PhysicalAttr, node, succ *PhysicalPlanNode) (bool, *PhysicalPlanNode) {
-	i := indexOfNode(node, succ.Predecessors())
-	if _, ok := succ.requiredAttrs()[i][requiredAttr.Key()]; ok {
+func requiredBySuccessor(requiredAttr PhysicalAttr, node, succ Node) (bool, Node) {
+	psucc, ok := succ.(*PhysicalPlanNode)
+	if !ok {
+		return false, succ
+	}
+
+	i := indexOfNode(node, psucc.Predecessors())
+	if _, ok := psucc.requiredAttrs()[i][requiredAttr.Key()]; ok {
 		return true, succ
 	}
-	if succ.passesThroughAttr(requiredAttr.Key()) {
+	if psucc.passesThroughAttr(requiredAttr.Key()) {
 		if len(succ.Successors()) == 0 {
 			return false, nil
 		}
 		// If this node does not require the attribute itself but passes it along,
 		// see if any successors require it.
-		for _, ssucc := range succ.Successors() {
-			if reqd, n := requiredBySuccessor(requiredAttr, succ, ssucc.(*PhysicalPlanNode)); !reqd {
+		for _, ssucc := range psucc.Successors() {
+			if reqd, n := requiredBySuccessor(requiredAttr, succ, ssucc); !reqd {
```

Comment:
I am trying to grok this loop and recurse. 

At first it looked like we may be able to remove the loop over `p.Successors()`. However, I see can that being problematic if two successors are on different branches. It looks like we are checking for **any** successor to require the `requiredAttr` in a successor branch of passthrough zero or more successor nodes. It also looks like the comment above wants an **any** successor check.

It looks like this loop guard will enforce that **all** successor branches to require the `requiredAttr`. For this to work as described I think we need to return `true, n`  if `reqd` is true:

```golang
if reqd, n := requiredBySuccessor(requiredAttr, succ, ssucc); reqd {
     return true, n
}
```

Current behavior:

```mermaid
flowchart TD;
   A[ requiredBySuccessor for node and succ  where `requiredAttr = 'foo'`] --> B[succ.Successors => SsuccA,  SsuccB \n does not require and passesThrough 'foo' ];
   B ---> C[range over SsuccA, SsuccB];
   C --  check requiredBySuccessor --> D[SsuccA, does not require or passthrough 'foo'];
   D -- false, SsuccA --> C;
   C -- false, SsuccA --> A;
   C -- never checked--> E[SsuccB, does require 'foo'];
```

Reply:
It is supposed to be an *all* check. If a node has an attribute that must be required by successors, there must be a node that requires it on every successor branch.

The only use case for "successor must require" is for plans that have parallelism, eg:
```
source |> aggregateWindow |> merge
```
In this case `source` and `aggregateWindow` have the `parallelRun` attribute and the `merge` node requires it. In effect, we process 4 (or whatever) parallel streams in `source` and `aggregateWindow` and then `merge` has to bring them together into a single stream that may be returned to the client or processed further.

If `aggregateWindow` had another successor besides the `merge` that did not require the `parallelRun` attribute, that would be an indication that something went wrong with planning, because we need parallel streams to be merged into one at some point before data is returned to the client. In effect this guards from creating incorrect plans. Does that make sense?

I can definitely update the comment to be more clear.